crates/scroll/evm/src/base_fee.rs

```diff
@@ -0,0 +1,256 @@
+use alloy_eips::calc_next_block_base_fee;
+use alloy_primitives::U256;
+use reth_chainspec::EthChainSpec;
+use reth_scroll_chainspec::{ChainConfig, ScrollChainConfig};
+use reth_storage_api::{BaseFeeProvider, StorageProvider};
+use scroll_alloy_evm::curie::L1_GAS_PRICE_ORACLE_ADDRESS;
+use scroll_alloy_hardforks::ScrollHardforks;
+
+/// L1 gas price oracle base fee slot.
+pub const L1_BASE_FEE_SLOT: U256 = U256::from_limbs([1, 0, 0, 0]);
+
+/// Protocol-enforced maximum L2 base fee.
+pub const MAX_L2_BASE_FEE: U256 = U256::from_limbs([10_000_000_000, 0, 0, 0]);
+
+/// The base fee overhead slot.
+const L1_BASE_FEE_OVERHEAD_SLOT: U256 = U256::from_limbs([101, 0, 0, 0]);
+
+/// The default base fee overhead, in case the L2 system contract isn't deployed or
+/// initialized.
+pub const DEFAULT_L1_BASE_FEE_OVERHEAD: U256 = U256::from_limbs([15_680_000, 0, 0, 0]);
+
+/// The base fee scalar slot.
+const L1_BASE_FEE_SCALAR_SLOT: U256 = U256::from_limbs([102, 0, 0, 0]);
+
+/// The default scalar applied on the L1 base fee, in case the L2 system contract isn't deployed or
+/// initialized.
+pub const DEFAULT_L1_BASE_FEE_SCALAR: U256 = U256::from_limbs([34_000_000_000_000, 0, 0, 0]);
+
+/// The precision of the L1 base fee.
+pub const L1_BASE_FEE_PRECISION: U256 = U256::from_limbs([1_000_000_000_000_000_000, 0, 0, 0]);
+
+/// The initial base fee.
+const INITIAL_BASE_FEE: u64 = 10_000_000;
+
+/// The Scroll base fee provider implementation.
+#[derive(Clone, Debug, Default)]
+pub struct ScrollBaseFeeProvider<ChainSpec>(ChainSpec);
+
+impl<ChainSpec> ScrollBaseFeeProvider<ChainSpec> {
+    /// Returns a new instance of a [`ScrollBaseFeeProvider`].
+    pub const fn new(chain_spec: ChainSpec) -> Self {
+        Self(chain_spec)
+    }
+}
+
+impl<ChainSpec, P> BaseFeeProvider<P> for ScrollBaseFeeProvider<ChainSpec>
+where
+    ChainSpec: EthChainSpec + ScrollHardforks + ChainConfig<Config = ScrollChainConfig>,
+    P: StorageProvider,
+{
+    fn next_block_base_fee<H: BlockHeader>(
+        &self,
+        provider: &mut P,
+        parent_header: &H,
+        ts: u64,
+    ) -> Result<u64, P::Error> {
+        let chain_spec = &self.0;
+
+        // load system contract into cache.
+        let system_contract_address = chain_spec.chain_config().l2_system_contract_address;
+        // query scalar and overhead.
+        let (mut scalar, mut overhead) = (
+            provider.storage(system_contract_address, L1_BASE_FEE_SCALAR_SLOT)?,
+            provider.storage(system_contract_address, L1_BASE_FEE_OVERHEAD_SLOT)?,
+        );
+        // if any value is 0, use the default values.
+        (scalar, overhead) = (
+            if scalar == U256::ZERO { DEFAULT_L1_BASE_FEE_SCALAR } else { scalar },
+            if overhead == U256::ZERO { DEFAULT_L1_BASE_FEE_OVERHEAD } else { overhead },
+        );
+
+        if chain_spec.is_feynman_active_at_timestamp(ts) {
+            Ok(feynman_base_fee(chain_spec, parent_header, ts, overhead.saturating_to()))
```

Comment:
Any risk about dropping back to `u64` here? I guess no since the result of `next_block_base_fee` must fit into `u64` anyway?

L2geth does not have this limitation, so if we set `overhead` incorrectly the behavior would diverge. But normally this value should be quite small.